### PR TITLE
*: Update to Prometheus 2.4.0

### PR DESCRIPTION
**Pull request by mxinden**
Supersedes https://github.com/coreos/prometheus-operator/pull/1894.

### CODE REVIEW

**Comment by mxinden** [comment]
Updating ` testAMZeroDowntimeRollingDeployment` to use Alertmanager v0.15.2 fails. An Alertmanager joining a given cluster does not properly catch up on which alerts have already been sent. I will debug this further upstream.

File: Documentation/compatibility.md [hunk 1]
```diff
@@ -35,3 +35,4 @@ The versions of Prometheus compatible to be run with the Prometheus Operator are
 * v2.2.1
 * v2.3.1
 * v2.3.2
+* v2.4.0
```
Comment (mrueg) [inline Documentation/compatibility.md]: Probably want to add 2.4.1 and 2.4.2 here as well

File: Documentation/user-guides/cluster-monitoring.md [hunk 2]
```diff
@@ -422,7 +422,7 @@ spec:
   serviceAccountName: prometheus-k8s
   serviceMonitorNamespaceSelector: {}
   serviceMonitorSelector: {}
-  version: v2.3.2
+  version: v2.4.0
```
Comment (mrueg) [inline Documentation/user-guides/cluster-monitoring.md]: Use 2.4.2 here

File: contrib/kube-prometheus/manifests/prometheus-prometheus.yaml [hunk 3]
```diff
@@ -25,4 +25,4 @@ spec:
   serviceAccountName: prometheus-k8s
   serviceMonitorNamespaceSelector: {}
   serviceMonitorSelector: {}
-  version: v2.3.2
+  version: v2.4.0
```
Comment (mrueg) [inline contrib/kube-prometheus/manifests/prometheus-prometheus.yaml]: Use 2.4.2 here

File: Documentation/compatibility.md (hunk 1 above)
Comment (deejay1) [reply Documentation/compatibility.md]: I'd stick just with 2.4.2 due to `[BUGFIX] Handle WAL corruptions properly prometheus/tsdb#389`

Comment (metalmatze) [reply Documentation/compatibility.md]: 2.4.2 is the new default Prometheus version, this is just the compatibility matrix of supported versions by the operator.

**Comment by metalmatze** [comment]
I think the e2e tests are still failing. IIRC @mxinden wanted to look into these, but I might be wrong here.

**Comment by mxinden** [comment]
@brancz @metalmatze Preparing a rebase and re-generate locally right now. Looking into the e2e failures afterwards.

**Comment by mxinden** [comment]
I can reproduce `TestAllNS/x/PromReloadConfig` failure locally.

**Comment by mxinden** [comment]
@brancz @metalmatze would you mind giving this another look?

File: test/framework/alertmanager.go [hunk 4]
```diff
@@ -255,6 +256,27 @@ func (f *Framework) CreateSilence(ns, n string) (string, error) {
 	return createSilenceResponse.Data.SilenceID, nil
 }
 
+func (f *Framework) SendAlertToAlertmanager(ns, n string) error {
+	var postAlertResp amAPIPostAlertResp
+	request := ProxyPostPod(f.KubeClient, ns, n, "web", "api/v1/alerts", `
+[{"labels":{"alertname":"ExampleAlert","prometheus":"my-prometheus"},"annotations":{},"startsAt":"2018-09-17T16:27:45.70751247Z","endsAt":"2030-09-17T16:31:45.80751247Z","generatorURL":"http://prometheus-test-0:9090/graph?g0.expr=vector%281%29\u0026g0.tab=1"}]
```
Comment (brancz) [inline test/framework/alertmanager.go]: let's not hardcode the dates here, even if it's in 12 years

File: test/e2e/alertmanager_test.go [hunk 5]
```diff
@@ -472,32 +470,43 @@ inhibit_rules:
...
+		go func() {
+			ticker := time.NewTicker(100 * time.Millisecond)
+			failures := 0
+			for {
+				select {
+				case <-ticker.C:
+					err := framework.SendAlertToAlertmanager(ns, "alertmanager-rolling-deploy-"+strconv.Itoa(replica))
+					if err != nil {
+						failures++
+						// Allow 30 (~3 Seconds) failures during Alertmanager rolling update.
```
Comment (brancz) [inline test/e2e/alertmanager_test.go]: actually any amount of failures is fine as long as the result in the end is still correct

Comment (mxinden) [reply test/e2e/alertmanager_test.go]: The test can be split up into three phases:

1. Send alerts and expect a single notification.

2. Roll out new version of Alertmanager.

3. Send alerts and expect no new notifications.

If sending alerts between two and three fails, we will not see any new notifications, thereby the test might give a false positive.

Does that make sense @brancz?

Comment (brancz) [reply test/e2e/alertmanager_test.go]: yep, makes perfect sense

File: test/framework/alertmanager.go (hunk 4 above)
Comment (mxinden) [reply test/framework/alertmanager.go]: Agree, that approach is a bit hacky. Updated.

**Comment by brancz** [comment]
lgtm :+1: nice!

**Pull request merged**